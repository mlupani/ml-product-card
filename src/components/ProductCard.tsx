import React, { createContext } from 'react';
import useProduct from '../hooks/useProduct';
import styles from '../styles/styles.module.css';
import { InitialValues, onChangeArgs, ProductCardContext, Producto, ProductCardHandlers } from '../interfaces/ProductCard';

export const ProductContext = createContext({} as ProductCardContext);
const { Provider } = ProductContext;

export interface Props {
    //children: React.ReactElement | React.ReactElement[];
    children: (args: ProductCardHandlers) => JSX.Element
    producto: Producto,
    className?: string,
    style?: React.CSSProperties,
    onChange?: (args: onChangeArgs) => void,
    value?: number
    initialValues?: InitialValues
}

export const ProductCard = ({ children, producto, className, style, onChange, value, initialValues }: Props) => {

    const { counter, increaseBy, reset, isMaxReached } = useProduct({ producto, onChange, value, initialValues });

    return (
        <Provider value={{ counter, increaseBy, producto, initialValues }}>
            <div className={`${styles.productCard} ${className}`} style={style}>
                {
                    children({
                        producto,
                        count: counter,
                        maxCount: initialValues?.maxCount,
                        increaseBy,
                        reset,
                        isMaxReached
                    })
                }
            </div>
        </Provider>
    )
}

export default ProductCard
