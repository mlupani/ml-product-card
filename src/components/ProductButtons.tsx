import React, { useCallback, useContext } from "react"
import { ProductContext } from "./ProductCard"
import styles from '../styles/styles.module.css';

export interface Props {
    className?: string,
    style?: React.CSSProperties
}

export const ProductButtons = ({className, style}: Props) => {

    const { increaseBy, counter, initialValues } = useContext(ProductContext)
    const isMaxReached = useCallback(
      () => !!initialValues?.maxCount && counter === initialValues.maxCount,
      [counter, initialValues?.maxCount],
    )

    return (
        <div className={`${styles.buttonsContainer} ${className}`} style={style}>
            <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>-</button>
            <div className={styles.countLabel}>{counter}</div>
            <button disabled={isMaxReached() ? true : false} className={`${styles.buttonAdd} ${isMaxReached() ? styles.disabled : null}`} onClick={() => increaseBy(+1)}>+</button>
        </div>
    )
}