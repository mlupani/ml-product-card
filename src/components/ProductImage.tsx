import React, { useContext } from "react";
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export interface Props {
    img?: string;
    className?: string;
    style?: React.CSSProperties;
}

export const ProductImage = ({img = '', className, style}: Props) => {

    const { producto } = useContext(ProductContext);
    let imageToShow;

    if (img) {
        imageToShow = img;
    } else if (producto.img) {
        imageToShow = producto.img;
    } else {
        imageToShow = noImage;
    }

    return (
        <img className={`${styles.productImg} ${className}`} src={imageToShow} alt="Product" style={style} />
    )
}