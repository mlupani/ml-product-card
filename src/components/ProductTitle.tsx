import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';

export interface Props {
    className?: string;
    title?: string;
    style?: React.CSSProperties;
}

export const ProductTitle = ({title, className= '', style} : Props) => {

    const { producto } = useContext(ProductContext);
    let titleToShow = title ? title : producto.title;

    return (
        <span className={`${styles.productDescription} ${className}`} style={style}>{titleToShow}</span>
    )
}