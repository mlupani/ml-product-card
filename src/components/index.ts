import { ProductTitle } from './ProductTitle';
import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';
import { ProductCardHOCProps } from '../interfaces/ProductCard';

export { ProductButtons } from "./ProductButtons";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";

export const ProductCard:ProductCardHOCProps = Object.assign(ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
});