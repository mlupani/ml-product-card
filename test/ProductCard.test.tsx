import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../src/components';
import { product1 } from './data/products';

const { act } = renderer;

describe('productCard', () => {

  it('render productCard with product title', () => {
    const wrapper = renderer.create(<ProductCard producto={product1}>{() => <h1>Producto</h1>}</ProductCard>).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it('test funcionality of counter' , () => {
    const wrapper = renderer.create(
        <ProductCard producto={product1}>
            {
                ({count, increaseBy}) => (
                    <>
                        <h1>Producto</h1>
                        <span>{count}</span>
                        <button onClick={() => increaseBy(1)}></button>
                    </>
                )
            }
        </ProductCard>);
        let tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
        act(() => {
            (tree as any).children[2].props.onClick()
        });
        tree = wrapper.toJSON();
        expect((tree as any).children[1].children[0]).toBe('1')

  })
});
