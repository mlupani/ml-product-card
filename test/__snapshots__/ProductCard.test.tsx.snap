// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`productCard render productCard with product title 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Producto
  </h1>
</div>
`;

exports[`productCard test funcionality of counter 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Producto
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;
